src/cascadia/TerminalSettingsModel/ActionArgsMagic.h

```diff
@@ -0,0 +1,160 @@
+// `true` in the definition of Equals() below, that's to deal with trailing
+// commas
+#define EQUALS_ARGS(type, name, jsonKey, required, ...) \
+    &&(otherAsUs->_##name == _##name)
+
+// JSON deserialization. If the parameter is required to pass any validation,
+// add that as the `required` parameter here, as the body of a conditional
+// EX: For the RESIZE_PANE_ARGS
+//    X(Model::ResizeDirection, ResizeDirection, "direction", args->ResizeDirection() == ResizeDirection::None, Model::ResizeDirection::None)
+// the bit
+//    args->ResizeDirection() == ResizeDirection::None
+// is used as the conditional for the validation here.
+#define FROM_JSON_ARGS(type, name, jsonKey, required, ...)                      \
+    if (required)                                                               \
+    {                                                                           \
+        return { nullptr, { SettingsLoadWarnings::MissingRequiredParameter } }; \
+    }                                                                           \
+    JsonUtils::GetValueForKey(json, jsonKey, args->_##name);
+
+// JSON serialization
+#define TO_JSON_ARGS(type, name, jsonKey, required, ...) \
+    JsonUtils::SetValueForKey(json, jsonKey, args->_##name);
+
+// Copy each property in the Copy() method
+#define COPY_ARGS(type, name, jsonKey, required, ...) \
+    copy->_##name = _##name;
+
+// hash each property in Hash(). You'll note there's a stray `0` in the
+// definition of Hash() below, that's to deal with trailing commas (or in this
+// case, leading.)
+#define HASH_ARGS(type, name, jsonKey, required, ...) \
+    , name()
+
+// Use ACTION_ARGS_STRUCT when you've got no other customizing to do.
+#define ACTION_ARGS_STRUCT(className, argsMacro)      \
+    struct className : public className##T<className> \
+    {                                                 \
+        ACTION_ARG_BODY(className, argsMacro)         \
+    };
+// Use ACTION_ARG_BODY when you've got some other methods to add to the args class.
+// case in point:
+//   * NewTerminalArgs has a ToCommandline method it needs to additionally declare.
+//   * GlobalSummonArgs has the QuakeModeFromJson helper
+
+#define ACTION_ARG_BODY(className, argsMacro)                                   \
+    className() = default;                                                      \
+    className(                                                                  \
+        argsMacro(CTOR_PARAMS) InitListPlaceholder = {}) :                      \
+        argsMacro(CTOR_INIT) _placeholder{} {};                                 \
+    argsMacro(DECLARE_ARGS);                                                    \
+                                                                                \
+private:                                                                        \
+    InitListPlaceholder _placeholder;                                           \

```

Comment:
You don't need to store the placeholder.

Reply:
`[[no_unique_address]]` will be helpful here in a few months with C++20.

Reply:
> few months

Wow you're optimistic, @lhecker.